test/e2e/cluster_upgrade_runtimesdk.go

```diff
@@ -266,26 +294,182 @@ func responsesConfigMap(name string, namespace *corev1.Namespace) *corev1.Config
+		if !clusterConditionShowsHookFailed(cluster, hookName) {
+			blocked = false
+		}
+		// Check if the Cluster is showing the RollingUpdateInProgress condition reason. If it has the update process is unblocked.
+		if conditions.IsFalse(cluster, clusterv1.ReadyCondition) &&
+			conditions.GetReason(cluster, clusterv1.ReadyCondition) == controlplanev1.RollingUpdateInProgressReason {
+			blocked = false
+		}
+		return blocked
+	}, intervals)
+}
+
+// afterControlPlaneUpgradeTestHandler provides an unblocked function which returns false if any MachineDeployment in the Cluster
+// has upgraded to the target Kubernetes version.
+func afterControlPlaneUpgradeTestHandler(ctx context.Context, c client.Client, namespace, clusterName, version string, intervals []interface{}) {
+	log.Logf("Blocking with AfterControlPlaneUpgrade hook")
+	hookName := "AfterControlPlaneUpgrade"
+	runtimeHookTestHandler(ctx, c, namespace, clusterName, hookName, func() bool {
+		var blocked = true
+		cluster := &clusterv1.Cluster{}
+		Eventually(func() error {
+			return c.Get(ctx, client.ObjectKey{Namespace: namespace, Name: clusterName}, cluster)
+		}).Should(Succeed())
+
+		// Check if the TopologyReconciled condition message contains both the hook name and hookFailedMessage.
+		if !clusterConditionShowsHookFailed(cluster, hookName) {
+			blocked = false
+		}
+
+		mds := &clusterv1.MachineDeploymentList{}
+		Eventually(func() error {
+			return c.List(ctx, mds, client.MatchingLabels{
+				clusterv1.ClusterLabelName:          clusterName,
+				clusterv1.ClusterTopologyOwnedLabel: "",
+			})
+		}).Should(Succeed())
+
+		// If any of the MachineDeployments have the target Kubernetes Version, the hook is unblocked.
+		for _, md := range mds.Items {
+			if *md.Spec.Template.Spec.Version == version {
+				blocked = false
+			}
+		}
+		return blocked
+	}, intervals)
+}
+
+// runtimeHookTestHandler runs a series of tests in sequence to check if the runtimeHook passed to is succeeds.
+//	  1) Checks that the hook has been called at least once.
+//    2) Check that the hook's blockingCondition is consistently true.
+//    - At this point the function sets the hook's response to be non-blocking.
+//    3) Check that the hook's blocking condition becomes true.
+// Note: runtimeHookTestHandler  assumes that the hook passed to it is currently returning a blocking response.
+// Updating the response to be non-blocking happens inline in the function.
+func runtimeHookTestHandler(ctx context.Context, c client.Client, namespace, clusterName, hookName string, blockingCondition func() bool, intervals []interface{}) {
+	// First check that the LifecycleHook has been called at least once.
+	Eventually(func() error {
+		return checkLifecycleHooksCalledAtLeastOnce(ctx, c, namespace, clusterName, []string{hookName})
+	}, 60*time.Second).Should(Succeed(), "%s has not been called", hookName)
+
+	// blockingCondition should consistently be true as the Runtime hook is returning "Failure".
```

Comment:
Q: Do we have a race condition between the hook has returned a blocking response + the blockingCondition is already true?

Let's assume the hook returned failure but the reconciler didn't set the condition yet. Will consistently get a false on the first run?


(Could be potentially mitigated if we have to by waiting for the condition in the Eventually in 440)